File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        counts = defaultdict(int)
        counts[0] = 1
        res = 0
        cur = 0

        for n in nums:
            cur += n
            res += counts[cur - goal]
            counts[cur] += 1
        return res

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        unordered_map<int, int> mp;
        mp[0] = 1;
        int res = 0, curr = 0;
        for (int n : nums) {
            curr += n;
            res += mp[curr - goal];
            mp[curr]++;
        }
        return res;
    }
};